util/dvsim/Deploy.py

```diff
@@ -279,118 +277,138 @@ def dispatch_cmd(self):
+                        msg = "```\n{}\n```\n".format(_get_n_lines(cnt, 5))
+                        self.fail_msg += msg
+                        log.log(VERBOSE, msg)
+                        failed = True
+                        chk_failed = False
+
+                if chk_passed:
+                    pattern = _find_patterns(self.pass_patterns, line)
+                    if pattern is not None:
+                        passed[pattern] = True
+                        chk_passed = not all(passed.values())
+
+        # If failed, then nothing else to do. Just return.
+        if failed:
             return False
 
         # If no fail patterns were seen, but the job returned with non-zero
         # exit code for whatever reason, then show the last 10 lines of the log
         # as the failure message, which might help with the debug.
         if self.process.returncode != 0:
-            msg = "Last 10 lines of the log:<br>\n"
+            msg = "Last 10 lines:\n```\n{}\n```\n".format(''.join(
+                lines[-10:]).strip())
             self.fail_msg += msg
             log.log(VERBOSE, msg)
-            get_fail_msg_cmd = "tail -n 10 " + self.log
-            msg = run_cmd(get_fail_msg_cmd)
-            msg = "```\n{}\n```\n".format(msg)
+            return False
+
+        # Ensure all pass patterns were seen.
+        if passed and chk_passed:
+            msg = "One or more pass patterns not found:\n{!r}\n".format(passed)
+            msg += "Last 10 lines:\n```\n{}\n```\n".format(''.join(
+                lines[-10:]).strip())
             self.fail_msg += msg
             log.log(VERBOSE, msg)
             return False
 
-        # If we get here, we've not seen anything explicitly wrong, but we
-        # might have "pass patterns": patterns that must occur in the log for
-        # the run to be considered successful.
-        for pass_pattern in self.pass_patterns:
-            grep_cmd = "grep -c -m 1 -E \'" + pass_pattern + "\' " + self.log
-            (status, rslt) = subprocess.getstatusoutput(grep_cmd)
-            if rslt == "0":
-                msg = "Pass pattern {!r} not found.<br>\n".format(pass_pattern)
-                self.fail_msg += msg
-                log.log(VERBOSE, msg)
-                return False
-
         return True
 
     def _link_odir(self, status):
-        old_link = self.sim_cfg.links['D'] + "/" + self.odir_ln
-        new_link = self.sim_cfg.links[status] + "/" + self.odir_ln
-        cmd = "ln -s " + self.odir + " " + new_link + "; "
-        cmd += "rm " + old_link
-        if os.system(cmd):
-            log.error("Cmd \"%s\" could not be run", cmd)
+        '''Soft-links the job's directory based on job's status, into
+        dispatched, running, passed, failed or killed directories in the
+        scratch area.'''
+
+        dest = Path(self.sim_cfg.links[status], self.odir_ln)
+        if os.path.exists(dest):
+            os.remove(dest)
+        os.symlink(self.odir, dest)
```

Comment:
There are race conditions here. See https://stackoverflow.com/questions/8299386/modifying-a-symlink-in-python#27788271 for how to do this properly. Also (although this *would* be racy) I wonder whether we should also be checking that `dest` was a symlink before?

Reply:
Well, you're right that we shouldn't run two copies of `dvsim` at the same time, but there's a standard atomic way to update a symlink, so I'd generally suggest just doing it properly.

Reply:
Done (in the other PR).